### ISSUE TITLE
Mind Vision

### ISSUE THREAD

**Issue by LordUsagi** [OregonCore/OregonCore#1212]
Mind Vision can't be cast over long distances currently. It also doesn't ignore LoS.

It should behave on a map per map basis, and be usable over any distance.

**Comment by adeutscher** [comment]
I started off with this by looking in `WorldSession::HandleFarSightOpcode`, since the first major related error message to pop up is "Player %s requests invalid seer".

When a user targets a player for Mind Vision:
1. `WorldSession::HandleFarSightOpcode` (MiscHandler.cpp,~line1271) detects that this is an application of Far Sight. Debug message prints "Added FarSight targetting of <target GUID> to to player <caster GUID>" (Clarification mine). This properly prints out GUID of the target, whether it be a player or an NPC.
2. `HandleFarSightOptCode` calls `Player::GetViewpoint`. If there's a farsight target, it will call `ObjectAccessor::GetObjectByTypeMask` to get more informations (Arguments are the casting player object, the GUID of the target, and the type mask (`TYPEMASK_SEER`).)
3. `ObjectAccessor::GetObjectByTypeMask` calls `GUID_HIPART`. This output spits out a value of 0 for players, which is in line with the `HIGHGUID_PLAYER` value of the `HighGuid` enum (defined in ObjectGuid.h).
4. `ObjectAccessor::GetObjectByTypeMask` switches off of the return value of `GUID_HIPART`, and matches to the case of `HIGHGUID_PLAYER`.
5. The check that controls whether or not to return the player `WorldObject` or fall through to return NULL is `if (typemask & TYPEMASK_PLAYER)`. This is where everything is goes sideways. The statement of `if (TYPEMASK_SEER & TYPEMASK_PLAYER` does not evaluate to true. When I cheat and hard-code this to true in my test build, then I'm able to follow my player target anywhere on the continent or if they go out of range (without a fix, going out of range would show the target running in place).

Double-checked by looking up the following type definitions out of ObjectGuid.h:

```
TYPEMASK_UNIT = 0x0008 // 8. Surprise!
TYPEMASK_DYNAMICOBJECT = 0x0040 // 64
TYPEMASK_SEER = TYPEMASK_UNIT | TYPEMASK_DYNAMICOBJECT // 72
TYPEMASK_PLAYER = 0x0010 // 16
```

The comment next to the definition for `TYPEMASK_UNIT` ("players also have it", Line 43 of ObjectGuid.h) suggests that `TYPEMASK_PLAYER` should be a combination of 0x0008 and `TYPEMASK_UNIT` in order to make sure that players count as units as far as the typemask is concerned. The alternative would be adding `TYPEMASK_PLAYER` onto `TYPEMASK_SEER`, but `TYPEMASK_UNIT`'s comment makes me want to go with the first option.

Changing the if statement to `if(typemask & (TYPEMASK_PLAYER | TYPEMASK_UNIT))` also worked to fix the issue. Stepping away for a bit while I build with the more long-term fix of changing `ObjectGuid.h`.

I found two other things that still affect Mind Vision. However, I believe that they might be a whole different can of worms. Worst-case, maybe something to be filed as a different issue? Problems:
- Mind Vision still allows itself to be cast across continents, though the debuff will never appear on the target. This being a problem fits in with the issue description and the tooltip.
- Refreshing Mind Vision on a faraway target (without stopping the original channel) will stop any updates of the remote site from being sent to the caster. Remote NPCs will stop patrolling after they reach their next waypoint, and remote players will stand or run in place. Targets that run near the caster again will be shown to move again, but when they run back out of range they will run in place similarly to how they would before the fix being discussed in this issue. For the moment, the problem looks like the refresh cast is never calling `WorldSession::HandleFarSightOpcode`. Haven't looked further up the chain yet.
